python/pfs/drp/stella/lsf.py

```diff
@@ -0,0 +1,688 @@
+from abc import ABC, abstractmethod
+from functools import partial
+import pickle
+import numpy as np
+from scipy.interpolate import interp1d
+from scipy.integrate import trapz
+
+from lsst.geom import Point2D, Point2I
+from lsst.afw.image import ImageF, ImageD
+from lsst.afw.math import FixedKernel
+from lsst.afw.geom.ellipses import Quadrupole
+
+__all__ = ["Kernel1D", "Lsf", "GaussianKernel1D", "GaussianLsf", "FixedEmpiricalLsf"]
+
+# Default interpolator factory
+DEFAULT_INTERPOLATOR = partial(interp1d, kind="linear", bounds_error=False, fill_value=0, copy=True,
+                               assume_sorted=True)
+
+
+class Kernel1D:
+    """A one-dimensional kernel
+
+    A ``Kernel1D`` is like a one-dimensional version of an
+    `lsst.afw.image.Image` that has an ``xy0`` set so that ``0,0`` is at the
+    center of the image. Main operations are ``convolve`` to convolve an array
+    by a kernel, and ``toArray`` which inserts the kernel at any point in an
+    array.
+
+    Parameters
+    ----------
+    values : `numpy.ndarray`, floating-point
+        Kernel values.
+    center : `int`
+        Index of center of kernel in the ``values``. It needn't be the middle.
+    normalize : `bool`, optional
+        Normalize the kernel to unity?
+    """
+    def __init__(self, values, center, normalize=True):
+        self.length = len(values)
+        self.center = center
+        self.values = values
+        self.indices = np.arange(self.length) - center
+        self.min = self.indices[0]
+        self.max = self.indices[-1]
+        if normalize:
+            self.values /= self.normalization()
+
+    @classmethod
+    def makeEmpty(cls, halfSize):
+        """Create an empty ``Kernel1D``
+
+        Parameters
+        ----------
+        halfSize : `int`
+            Half-size of the kernel.
+
+        Returns
+        -------
+        self : ``cls``
+            Empty kernel.
+        """
+        size = 2*halfSize + 1
+        return cls(np.zeros(size, dtype=float), halfSize, normalize=False)
+
+    def __len__(self):
+        """Number of elements in kernel array"""
+        return self.length
+
+    def _indexOutOfRange(self, index):
+        """Is the index out of range?"""
+        return index < self.min or index > self.max
+
+    def _getitem(self, index):
+        """Get a single value from the kernel
+
+        Values off the end are zero.
+        """
+        return self.values[index + self.center] if not self._indexOutOfRange(index) else 0.0
+
+    def _setitem(self, index, value):
```

Comment:
Isn't this a `__` method, solely used by `__setitem__`?
And again, aren't we doing the index check twice, once here, then in the index operation?  I assume that you're trying to provide better error messages, but we could do that in a `except` block

Reply:
Leading double underscore is [used to prevent the method from being overridden by subclasses](http://igorsobreira.com/2010/09/16/difference-between-one-underline-and-two-underlines-in-python.html), so I don't think it's appropriate here.
But I've reworked the getter and setter by using `np.broadcast_arrays` instead of `np.vectorize`; that should be faster, and it eliminates the confusing `_getitem` and `_setitem` methods.